File: client/local-server-plugin/ios/Sources/LocalServerPlugin/LocalServer.swift
import Foundation
import Network

// LAN transport core for iOS (poker_app issue #28).
//
// Design (Network.framework, iOS 13+):
//   - advertise + listen: NWListener with an NWListener.Service for Bonjour
//     (_feltpoker._tcp) and the WebSocket protocol option, so every peer hosts
//     its own endpoint. No privileged host, matching the mesh design.
//   - browse: NWBrowser over the same service type emits peerDiscovered.
//   - connect: NWConnection to a discovered endpoint with .ws options.
//
// The structure below is complete and compilable in intent; the marked bodies
// are stubs because they need a real device / two devices on a LAN to validate,
// which cannot be done from this environment. Each TODO is precise.

final class LocalServer {
    typealias OnMessage = (_ peerId: String, _ data: String) -> Void
    typealias OnConnection = (_ peerId: String, _ state: String) -> Void
    typealias OnDiscovered = (_ peer: [String: Any]) -> Void

    private let onMessage: OnMessage
    private let onConnection: OnConnection
    private let onDiscovered: OnDiscovered

    private var listener: NWListener?
    private var browser: NWBrowser?
    private var connections: [String: NWConnection] = [:]
    private var selfPeerId: String = ""
    private let queue = DispatchQueue(label: "feltpoker.localserver")

    init(onMessage: @escaping OnMessage, onConnection: @escaping OnConnection, onDiscovered: @escaping OnDiscovered) {
        self.onMessage = onMessage
        self.onConnection = onConnection
        self.onDiscovered = onDiscovered
    }

    func start(peerId: String, serviceType: String, port: UInt16, displayName: String) throws -> UInt16 {
        selfPeerId = peerId
        // TODO(device): build NWParameters with .tcp + NWProtocolWebSocket.Options,
        // create NWListener(using:on:), set listener.service =
        // NWListener.Service(name: displayName, type: serviceType), set
        // newConnectionHandler to accept peers and pump receiveMessage into
        // onMessage, then listener.start(queue:). Also start browse().
        // Return listener.port?.rawValue once bound (0 -> OS-assigned).
        startBrowse(serviceType: serviceType)
        return port
    }

    private func startBrowse(serviceType: String) {
        // TODO(device): NWBrowser(for: .bonjour(type: serviceType, domain: nil),
        // using: .init()); on browseResultsChangedHandler, resolve endpoints and
        // call onDiscovered(["peerId": ..., "host": ..., "port": ...]). The peer
        // id is exchanged as a first WS frame on connect (endpoints only carry
        // the Bonjour name).
    }

    func connect(peerId: String, host: String, port: UInt16) {
        guard connections[peerId] == nil else { return }
        // TODO(device): NWConnection(host:port:using: ws params); on .ready send a
        // hello frame carrying selfPeerId, then onConnection(peerId, "open"); pump
        // receiveMessage into onMessage; on failed/cancelled onConnection(peerId,
        // "closed"). Store in connections[peerId].
    }

    func send(peerId: String, data: String) {
        guard let conn = connections[peerId] else { return }
        let metadata = NWProtocolWebSocket.Metadata(opcode: .text)
        let context = NWConnection.ContentContext(identifier: "text", metadata: [metadata])
        conn.send(content: data.data(using: .utf8), contentContext: context, completion: .contentProcessed { _ in })
    }

    func disconnect(peerId: String) {
        connections[peerId]?.cancel()
        connections[peerId] = nil
        onConnection(peerId, "closed")
    }

    func stop() {
        listener?.cancel()
        browser?.cancel()
        for (_, c) in connections { c.cancel() }
        connections.removeAll()
    }
}

File: client/local-server-plugin/android/src/main/java/com/feltpoker/localserver/LocalServer.java
package com.feltpoker.localserver;

import java.util.HashMap;
import java.util.Map;

// LAN transport core for Android (poker_app issue #28).
//
// Design:
//   - advertise + listen: NsdManager.registerService(_feltpoker._tcp) plus a
//     local WebSocket listener (NanoHTTPD's NanoWSD, or Java-WebSocket's
//     WebSocketServer) bound to the same port, so every peer hosts its own
//     endpoint (no privileged host).
//   - discover: NsdManager.discoverServices + resolveService emits onDiscovered.
//   - connect: a WebSocket client (Java-WebSocket) dials a resolved peer.
//
// The class shape is complete; the marked bodies are stubs because validating
// mDNS + sockets needs two devices on a Wi-Fi LAN. Each TODO is precise. Wire
// the concrete NsdManager/NanoWSD calls here; the plugin bridge above is done.

public class LocalServer {

    public interface OnMessage { void call(String peerId, String data); }
    public interface OnConnection { void call(String peerId, String state); }
    public interface OnDiscovered { void call(String peerId, String host, int port, String displayName); }

    private final OnMessage onMessage;
    private final OnConnection onConnection;
    private final OnDiscovered onDiscovered;

    private final Map<String, Object> connections = new HashMap<>();
    private String selfPeerId = "";
    private int boundPort = 0;

    public LocalServer(OnMessage onMessage, OnConnection onConnection, OnDiscovered onDiscovered) {
        this.onMessage = onMessage;
        this.onConnection = onConnection;
        this.onDiscovered = onDiscovered;
    }

    public int start(String peerId, String serviceType, int port, String displayName) throws Exception {
        this.selfPeerId = peerId;
        // TODO(device): start a NanoWSD/WebSocketServer bound to `port` (0 -> OS
        // picks; read the actual port into boundPort). In its onOpen, read the
        // peer's hello frame (its peerId) then onConnection(peerId, "open"); in
        // onMessage, forward to this.onMessage; onClose -> onConnection(peerId,
        // "closed"). Then registerService via NsdManager with serviceType and
        // displayName, and startDiscovery (see startDiscovery()).
        boundPort = port;
        startDiscovery(serviceType);
        return boundPort;
    }

    private void startDiscovery(String serviceType) {
        // TODO(device): NsdManager.discoverServices(serviceType, PROTOCOL_DNS_SD,
        // listener); on service found, resolveService; on resolved, call
        // onDiscovered(name, host.getHostAddress(), port, name). Skip our own
        // advertised service (name == selfPeerId's display name).
    }

    public void connect(String peerId, String host, int port) {
        if (connections.containsKey(peerId)) return;
        // TODO(device): open a Java-WebSocket client to ws://host:port; on open,
        // send a hello frame carrying selfPeerId, then onConnection(peerId,
        // "open"); on message forward to onMessage; on close onConnection(peerId,
        // "closed"). Store the client in connections.
    }

    public void send(String peerId, String data) {
        Object conn = connections.get(peerId);
        if (conn == null) return;
        // TODO(device): ((WebSocket) conn).send(data);
    }

    public void disconnect(String peerId) {
        connections.remove(peerId);
        onConnection.call(peerId, "closed");
        // TODO(device): close the stored socket.
    }

    public void stop() {
        connections.clear();
        // TODO(device): unregisterService, stopDiscovery, stop the WS server.
    }
}
